File: aws-globalaccelerator-endpointgroup/src/main/java/software/amazon/globalaccelerator/endpointgroup/ReadHandler.java
package software.amazon.globalaccelerator.endpointgroup;

import com.amazonaws.services.globalaccelerator.model.EndpointGroup;
import lombok.val;
import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy;
import software.amazon.cloudformation.proxy.HandlerErrorCode;
import software.amazon.cloudformation.proxy.Logger;
import software.amazon.cloudformation.proxy.ProgressEvent;
import software.amazon.cloudformation.proxy.OperationStatus;
import software.amazon.cloudformation.proxy.ResourceHandlerRequest;

public class ReadHandler extends BaseHandler<CallbackContext> {

    @Override
    public ProgressEvent<ResourceModel, CallbackContext> handleRequest(
        final AmazonWebServicesClientProxy proxy,
        final ResourceHandlerRequest<ResourceModel> request,
        final CallbackContext callbackContext,
        final Logger logger) {

        final ResourceModel model = request.getDesiredResourceState();

        val agaClient = AcceleratorClientBuilder.getClient();
        logger.log(String.format("Read request for endpoint group: [%s]", request));

        val endpointGroup = HandlerCommons.getEndpointGroup(model.getEndpointGroupArn(), proxy, agaClient, logger);
        val endpointGroupResourceModel = convertEndpointGroupToResourceModel(endpointGroup);

        if (endpointGroupResourceModel == null){
            logger.log(String.format("Endpoint group with ARN [%s] not found", model.getEndpointGroupArn()));
            return ProgressEvent.defaultFailureHandler(new Exception("Endpoint group not found."), HandlerErrorCode.NotFound);
        } else {
            return ProgressEvent.defaultSuccessHandler(endpointGroupResourceModel);
        }
    }

    private ResourceModel convertEndpointGroupToResourceModel(EndpointGroup endpointGroup) {
        ResourceModel newModel = null;
        if (endpointGroup != null) {
            newModel = new ResourceModel();
            newModel.setEndpointGroupArn(endpointGroup.getEndpointGroupArn());
            newModel.setHealthCheckIntervalSeconds(endpointGroup.getHealthCheckIntervalSeconds());
            newModel.setHealthCheckPath(endpointGroup.getHealthCheckPath());
            newModel.setHealthCheckPort(endpointGroup.getHealthCheckPort());
            newModel.setHealthCheckProtocol(endpointGroup.getHealthCheckProtocol());
            newModel.setThresholdCount(endpointGroup.getThresholdCount());
            newModel.setTrafficDialPercentage(endpointGroup.getTrafficDialPercentage().intValue());
            newModel.setEndpointGroupRegion(endpointGroup.getEndpointGroupRegion());
        }
        return newModel;
    }
}

File: aws-globalaccelerator-endpointgroup/src/main/kotlin/software/amazon/globalaccelerator/endpointgroup/ReadHandler.kt
package software.amazon.globalaccelerator.endpointgroup

import com.amazonaws.services.globalaccelerator.model.EndpointDescription
import com.amazonaws.services.globalaccelerator.model.EndpointGroup
import com.amazonaws.services.globalaccelerator.model.PortOverride
import software.amazon.cloudformation.proxy.AmazonWebServicesClientProxy
import software.amazon.cloudformation.proxy.HandlerErrorCode
import software.amazon.cloudformation.proxy.Logger
import software.amazon.cloudformation.proxy.ProgressEvent
import software.amazon.cloudformation.proxy.ResourceHandlerRequest

/**
 * Read handler implementation for Endpoint Group resource.
 */
class ReadHandler : BaseHandler<CallbackContext>() {
    override fun handleRequest(proxy: AmazonWebServicesClientProxy,
                               request: ResourceHandlerRequest<ResourceModel>,
                               callbackContext: CallbackContext?,
                               logger: Logger): ProgressEvent<ResourceModel, CallbackContext?> {
        logger.debug("Read EndpointGroup request: $request")
        val model = request.desiredResourceState
        val agaClient = AcceleratorClientBuilder.client
        val endpointGroup = HandlerCommons.getEndpointGroup(model.endpointGroupArn, proxy, agaClient, logger)
        val endpointGroupResourceModel = convertEndpointGroupToResourceModel(endpointGroup)
        return if (endpointGroupResourceModel == null) {
            logger.debug("Endpoint group with ARN [${model.endpointGroupArn}] not found")
            ProgressEvent.defaultFailureHandler(Exception("Endpoint group not found."), HandlerErrorCode.NotFound)
        } else {
            ProgressEvent.defaultSuccessHandler(endpointGroupResourceModel)
        }
    }

    private fun convertEndpointGroupToResourceModel(endpointGroup: EndpointGroup?): ResourceModel? {
        return if (endpointGroup != null) {
            ResourceModel().apply {
                this.endpointGroupArn = endpointGroup.endpointGroupArn
                this.healthCheckIntervalSeconds = endpointGroup.healthCheckIntervalSeconds
                this.healthCheckPath = endpointGroup.healthCheckPath
                this.healthCheckPort = endpointGroup.healthCheckPort
                this.healthCheckProtocol = endpointGroup.healthCheckProtocol
                this.thresholdCount = endpointGroup.thresholdCount
                this.trafficDialPercentage = endpointGroup.trafficDialPercentage.toDouble()
                this.endpointGroupRegion = endpointGroup.endpointGroupRegion
                this.endpointConfigurations = getEndpointConfigurations(endpointGroup.endpointDescriptions)
                this.portOverrides = getPortOverrides(endpointGroup.portOverrides)
            }
        } else {
            null
        }
    }
}
